_posts/2021-08-29-mosaic-fusion.md

```diff
@@ -0,0 +1,100 @@
+positioning and alignment of the individual image tiles is already available. In our case,
+this information is available based on the the metadata recorded by the microscope, however this
+information could also be from  previous registration step that matches image features 
+in the overlapping areas.
+ 
+
+## The solution
+
+
+The array that can hold resulting mosaic image will often have a size that is too large 
+to fit in RAM, therefore we will use dask arrays and the `map_block` function to enable 
+out-of-core processing. As an added benefit, we will get parallel processing for free
+which speeds up the fusion process.
+
+Typically whenever we want to join dask arrays, we use [Stack, Concatenate, and Block](https://docs.dask.org/en/latest/array-stack.html). However, these are not good tools for mosaic image fusion, because:
+
+1. the image tiles will be be overlapping,
+2. in the general case, individual image tiles may not be positioned on an exact grid and they can also have small rotations.
+
+Using the `block_info` dictionary and creating some pseudo-code together, Volker managed to implement mosaic fusion with `dask-image`. From the [Dask documentaion](https://docs.dask.org/en/latest/generated/dask.array.map_blocks.html?highlight=block_info#dask.array.map_blocks):
+> Your block function gets information about where it is in the array by accepting a special `block_info` or `block_id` keyword argument.
+
+The basic outline of the image mosiac workflow is as follows:
+1. Read in image tiles acquired from the microscope.
+2. Calculate the location of the image tiles, using information from the sample stage position at the time each image tile was acquired.
+3. Use an affine transformation to place the image tiles in their proper location.
+4. Fuse the combined images into a single result.
+
+
+### Results
+
+Using the dask-based image fusion method, we could achieve a speed up from several hours to tens of minutes compared to 
+a previous method based on an ImageJ plugin for datasets with many image tiles (~500-1000 tiles) on similar workstation hardware.
+Due to dask's ability to handle data out-of-core and chunked array storage using zarr it is also possible to run the 
+fusion on hardware with limited RAM.
+
+The image below shows mosaic image fusion for a smaall examples with 63 image tiles, displayed in the napari image viewer.
+
+![Mosaic fusion images in the napari image viewer](/images/mosaic-fusion/NapariMosaics.png)
+
+And here is an animation of whole slide mosaic fusion.
+
+![Animation of whole slide mosaic fusion images](/images/mosaic-fusion/Lama_whole_slide.gif)
+
+### Code
+
+Code relatiing to this mosaic image fusion project can be found in the `DaskFusion` GitHub repository here: https://github.com/VolkerH/DaskFusion
+
+There is a self-contained example available in [this notebook](https://github.com/VolkerH/DaskFusion/blob/main/Load_Mosaic.ipynb), which downloads reduced-size example data to demonstrate the process.
```

Comment:
```suggestion
There is a self-contained example available in [this notebook](https://github.com/VolkerH/DaskFusion/blob/main/DaskFusion_Example.ipynb), which downloads reduced-size example data to demonstrate the process.
```

Notebook has been renamed (not yet merged, but we can assume it will be by the time this is published)